### PR TITLE
Add guide for push notifications

### PR DESCRIPTION
**Pull request by ChristianPauly**
I hope that this can be useful for other app developers and there are no more typos in there. English is not my native language.

### CODE REVIEW

File: appdev/guides/pushnotifications.rst [hunk 1]
```diff
@@ -0,0 +1,235 @@
+Push Notifications
+==================
+
+Let's assume that you have an app created with Clickable and published on the OpenStore. But now you want to be able to send Push Notifications to your users. First of all, you need to understand how this is working:
+
+.. image:: pushoverview.jpg
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: You have excellent handwriting and this chart makes sense visually. However, I would prefer that charts in the documentation are made using a computer graphics tool such as https://draw.io.

File: appdev/guides/pushnotifications.rst [hunk 2]
```diff
@@ -0,0 +1,235 @@
+Push Notifications
+==================
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: This document has all of the content I need to start using push notifications in my app. However, the layout of the content makes it feel a bit hard to follow. It is trying to be two documents, I feel:

1. How to set an app up for receiving push notifications
1. What to do with a notification when you get it

My gut feeling is that these two topics are logically separated and each requires its own section, if not its own document. What do you think?

Comment (UniversalSuperBox) [reply appdev/guides/pushnotifications.rst]: To add to that, the push helper can grab notifications independently of the push service, right? That makes all the more reason to have two separate documents... maybe even a chapter of documentation dedicated to notifications.

File: appdev/guides/pushnotifications.rst (hunk 1 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: You are right of course. :D I will create one

Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: Its done now =)

File: appdev/guides/pushnotifications.rst (hunk 2 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: What do you think about the latest commit with "Restructuring"? It is no divided into two topics:
"Make your app ready for Push Notifications"
"Using the push notifications API"

File: appdev/guides/pushnotifications.rst [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+.. note::
+    In the following example we will not implement a server. Also the communication between your app and your server is up to you. Please inform the user about the communication with your server by providing a privacy policy!
+
+Make the app ready for push notifications
+-----------------------------------------
+
+Implementing the PushClient
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+First we need to add the policy group „push-notification-client“. Your apparmor file could look like this::
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: Please replace smart quotes with non-smart quotes (`"`)

File: appdev/guides/pushnotifications.rst [hunk 4]
```diff
@@ -0,0 +1,235 @@
+Push notifications
+==================
+
+Let's assume that you have an app created with Clickable and published on the OpenStore. But now you want to be able to send Push Notifications to your users. First of all, you need to understand how this is working:
+
+.. image:: ../../_static/images/appdev/guides/pushnotifications/Diagram.png
+
+1. Every app which uses push notifications has got a unique token. This token identifies the user, the device and the app itself. The token is generated by the UBports Push Service.
+2. You will need the token to send a push notification so the app needs to send this token to the entity, which will send the notification.
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: Would this be clearer as "the app sends its token to the app developer's server"?

File: appdev/guides/pushnotifications.rst [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+3. With the token you can send a simple HTTP request to the UBports Push Server which will forward the notification the user's device.
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: As a style change, I prefer to not use "just," "simple," "easy," or "clear" when referring to actions taken from our documentation. They might not be easy or clear for some people.

* https://dev.to/marcuscreo/the-4-letter-word-word-that-makes-my-blood-boil
* https://www.glamour.com/story/why-i-quit-using-the-word-just-in-my-emails

That said, I need to put in a PR to remove it from other places.

File: appdev/guides/pushnotifications.rst [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+Implementing the PushClient
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+First we need to add the policy group „push-notification-client“. Your apparmor file could look like this::
+
+	{
+	    "policy_groups": [
+		"networking",
+		"push-notification-client"
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: In code examples, it may be helpful to highlight certain lines. That could be useful here to show which line was added.

To do this, use the `code-block` directive rather than `::`. Then, add the `:emphasize-lines:` argument to the directive like this:

```
.. code-block:: js
   :emphasize-lines: 3

    ... some code ...
```

http://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block

File: appdev/guides/pushnotifications.rst [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+	[…]
+
+	PushClient {
+		id: pushClient
+		appId: "appname.yourname_hookname"
+		onTokenChanged: console.log("👍", pushClient.token)
+	}
+
+You need to set the correct appId! If the app name in your manifest file is appname.yourname and the name of the hook is hookname, then the appId is:  appname.yourname_hookname.
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: You're mentioning the hook name before you've set up the push helper.

File: appdev/guides/pushnotifications.rst [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+When we now start the app, it will get a token and print this token in the logs. With clickable logs we will be able to copy this token out of the terminal. But the app is not yet ready to receive a push notification. For this we need something called a pushhelper!
+
+Implementing the push helper
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The pushhelper is a part of the app which will receive all push notifications and process them before sending them to the system notification center. It will receive a json-file and must outpot another json-file in the correct format. The pushhelper is seperated from the app. So we need a new hook in the manifest. It could look like this::
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: outpot -> output

File: appdev/guides/pushnotifications.rst (hunk 2 above)
Comment (UniversalSuperBox) [reply appdev/guides/pushnotifications.rst]: Much better!

File: appdev/guides/pushnotifications.rst [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+	    "policy_version": 16.04
+	}
+
+The push.json is just for redirecting to the executable file::
+
+	{
+	    "exec": "pushexec"
+	}
+
+In our tutorial we will use python to create a simple executable which will just forward the notification without chaning anything::
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: chaning -> changing

File: appdev/guides/pushnotifications.rst [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+	#!/usr/bin/python3
+
+	import sys
+
+	f1, f2 = sys.argv[1:3]
+
+	open(f2, "w").write(open(f1).read())
+
+If we don't have done it already we also need to add this new files to the cmakelist and make the pushexec executable::
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: Drop the start of this sentence so it becomes "We also need to add these new files..."

File: appdev/guides/pushnotifications.rst (hunk 4 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: You're right. I have changed it

File: appdev/guides/pushnotifications.rst (hunk 5 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: Okay, I have removed these words.

File: appdev/guides/pushnotifications.rst (hunk 7 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: That's the hookname of the app, not the hookname of the pushhelper. I could change it to:
appId: "appname.yourname_apphookname" to make it more clear?

File: appdev/guides/pushnotifications.rst (hunk 6 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: I've added some code highlighting and highlighted lines now

File: appdev/guides/pushnotifications.rst (hunk 10 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: Done :)

File: appdev/guides/pushnotifications.rst [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    In the following example we will not implement a server. Also the communication between your app and your server is up to you. Please inform the user about the communication with your server by providing a privacy policy!
+
+Make the app ready for push notifications
+-----------------------------------------
+
+Implementing the PushClient
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Implementing the pushclient
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: Repeated header

File: appdev/guides/pushnotifications.rst (hunk 6 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: Oops, sorry. Is fixed

File: appdev/guides/pushnotifications.rst (hunk 11 above)
Comment (ChristianPauly) [reply appdev/guides/pushnotifications.rst]: Don't know how this could happen... maybe a wrong copy&paste action. Sorry... this is fixed now

File: appdev/guides/pushnotifications.rst [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+	    "title": "myapp",
+	    "hooks": {
+		"myapp": {
+		    "apparmor": "myapp.apparmor",
+		    "desktop":  "myapp.desktop"
+		},
+		"push": {
+		    "apparmor": "push-apparmor.json",
+		    "push-helper": "push.json"
+		}
```
Comment (UniversalSuperBox) [inline appdev/guides/pushnotifications.rst]: Should lines 10-13 be highlighted instead of 12 and 13 here?

Also, is it strictly required that the myapp.apparmor has the push-notification-client if the push helper has it?
